### DISCUSSION TITLE
picolibc usage question

### DISCUSSION THREAD

**Discussion by e-sr** [picolibc/picolibc#474] [Q&A] [answered]
I have trouble understanding the usage of picolibc in startup code, understanding the linking process and gcc flags in general.

Working on a m4f with newlib normally i have a startup code which defines the interrupt vector and the reset vector handler where following tasks are done: 
copy and initailize memory, setup FPU, call `__init_array_start` and finally call `main`. I have not written this code, it comes from the sdk i am using (ti msp432E). The code is strongly coupled with memory definition in linker file and written to work with newlib.
Linking is done using flags as 

`-lnosys -nostartfiles -static -specs=nosys.specs -specs=nano.specs` (or newlib.specs)

Now changing to picolibc i dont' know exactly wich linker flags to use and specifically how to use the startup functions in crt0. 
I assume i have to call `__start()` somewhere in the resetISR, however i the hello-world example and anywere in the doc there is no startup code example  and no usage of `__start`. 
The example assumes using the provided linker script as "straightforward" solution. However `picolibc.ld` is, at lest for me, a pretty involved linker script. For example i have trouble to understand the section name for the interrupt vector, like this?
```c
void (* const interruptVectors[])(void) __attribute__ ((section (".init.enter"))) =...
```
and looking at the complexity it discourage me to try to write my own.

Another point wich is unclear is the docs say to uninstall/remove newlib(GCC9) because no flag `-nolibc` . On the other side,the doc says using mason you have to use `-lgcc  -nolibc`.

I am now using 
`-lnosys -specs=nosys.specs -nolibc -lgcc -specs=picolibc.specs`
is that correct?

As you see i'm not really familiar with the linking process and all details involved in linker flags usage but after searching the web, looking at gcc man...i apologize and i have to ask.

**Comment by keith-packard** [comment, marked as answer]
No need to apologize; this is the hardest part of building embedded software.

First off, the linker scripts and startup code provided with picolibc are provided for a couple of reasons:

 1. Enables stand-alone testing using qemu.
 2. Enables build-system compiler tests that require linking.
 3. Provides a sample for people getting started in this space
 4. Can be used in some cases for production code, if your project is simple enough.
 5. Used by my personal projects (including AltOS and Snek).

RTOS projects that include picolibc support, like FreeRTOS and Zephyr, do not use them. They already have a sophisticated build environment with adequate support for what picolibc needs. 

The sample startup code is mostly intended to demonstrate the steps needed to get the hardware initialized and running code. In particular, the interrupt vector setup for ARM devices is unlikely to match your target at all.

The sample linker script is complicated by including support for thread-local storage (TLS), needing to support multiple architectures and including some bits likely only used by my own personal projects.  It has some "clever hacks" for TLS support -- it creates a static initial TLS area right in the script which allows TLS to be used without requiring any allocation or setup beyond the usual .data/.bss code, other than setting the TLS base pointer to the correct value.

For your use, it sounds like you probably want to just start with the existing linker script and startup code and possibly modify them in small ways (if at all). If you disable TLS support while building picolibc (which is likely already happening as most embedded GCC toolchains seem to disable this for some reason), then the existing startup code and linker script are likely to 'just work'.

To make this work, I think you want the following flags:
```
gcc --specs=picolibc.specs -T<your linker script file name here> -nostartfiles <your startup file name here> $(OBJS)
```

The magic in `picolibc.specs` will find the correct C library and include `-lgcc` after that. The `-T` option that references your existing linker script will cause the compiler to *not* use the picolibc linker script. `-nostartfiles` means you won't use the picocrt bits from picolibc as you're using your own bits from the SDK.

Note that you don't want to use `-Wl,-T` instead of `-T` -- `picolibc.specs` only avoids using `picolibc.ld` when you use the `-T` form.
